modules/cudaarithm/src/cuda/transpose.cu

```diff
@@ -56,40 +56,136 @@ using namespace cv;
+    static void call(const cv::cuda::GpuMat& src, cv::cuda::GpuMat& dst, cudaStream_t stream)
     {
-        NppStreamHandler h(StreamAccessor::getStream(stream));
+      NppiSize srcsz;
+      srcsz.height = src.rows;
+      srcsz.width = src.cols;
 
-        NppiSize sz;
-        sz.width  = src.cols;
-        sz.height = src.rows;
+      cv::cuda::NppStreamHandler h(stream);
 
-        nppSafeCall( nppiTranspose_8u_C1R(src.ptr<Npp8u>(), static_cast<int>(src.step),
-            dst.ptr<Npp8u>(), static_cast<int>(dst.step), sz) );
+      nppSafeCall( func(src.ptr<npp_type>(), static_cast<int>(src.step), dst.ptr<npp_type>(), static_cast<int>(dst.step), srcsz) );
 
-        if (!stream)
-            CV_CUDEV_SAFE_CALL( cudaDeviceSynchronize() );
+      if (stream == 0)
+        cudaSafeCall( cudaDeviceSynchronize() );
     }
-    else if (elemSize == 4)
-    {
-        gridTranspose(globPtr<int>(src), globPtr<int>(dst), stream);
-    }
-    else // if (elemSize == 8)
+  };
+
+  #if USE_NPP_STREAM_CONTEXT
+  template <int DEPTH, typename NppTransposeFunc<DEPTH>::func_ctx_t func> struct NppTransposeCtx
+  {
+    typedef typename NppTransposeFunc<DEPTH>::npp_type npp_type;
+
+    static void call(const cv::cuda::GpuMat& src, cv::cuda::GpuMat& dst, cudaStream_t stream)
     {
-        gridTranspose(globPtr<double>(src), globPtr<double>(dst), stream);
+      NppiSize srcsz;
+      srcsz.height = src.rows;
+      srcsz.width = src.cols;
+
+      NppStreamContext ctx;
+      nppSafeCall( nppGetStreamContext(&ctx) );
+      ctx.hStream = stream;
+
+      nppSafeCall( func(src.ptr<npp_type>(), static_cast<int>(src.step), dst.ptr<npp_type>(), static_cast<int>(dst.step), srcsz, ctx) );
     }
+  };
+  #endif
+}
+
+void cv::cuda::transpose(InputArray _src, OutputArray _dst, Stream& stream)
+{
+    GpuMat src = getInputMat(_src, stream);
 
-    syncOutput(dst, _dst, stream);
+    const int srcType = src.type();
+    const size_t elemSize = src.elemSize();
+
+    GpuMat dst = getOutputMat(_dst, src.cols, src.rows, src.type(), stream);
+
+    const bool isSupported =
+      (elemSize == 1) || (elemSize == 2) || (elemSize == 3) || (elemSize == 4) ||
+      (elemSize == 6) || (elemSize == 8) || (elemSize == 12) || (elemSize == 16);
+
+    if (!isSupported)
+        CV_Error(Error::StsUnsupportedFormat, "");
+    else if (src.empty())
+        CV_Error(Error::StsBadArg, "image is empty");
+
+    if ((src.rows == 1) && (src.cols == 1))
+        src.copyTo(dst, stream);
+    else if (src.rows == 1)
+        src.reshape(0, src.cols).copyTo(dst, stream);
+    else if ((src.cols == 1) && src.isContinuous())
+        src.reshape(0, src.cols).copyTo(dst, stream);
+    else
+    {
+        #if USE_NPP_STREAM_CONTEXT
+        constexpr const bool useNppStreamCtx = true;
+        #else
+        constexpr const bool useNppStreamCtx = false;
+        #endif
+        cudaStream_t _stream = StreamAccessor::getStream(stream);
+
+        if (!_stream || !useNppStreamCtx)
```

Comment:
Would it be better to always use the old api when CUDA SDK < 10.1 and the newer one otherwise regardless of whether there is a stream or not?